### PR TITLE
Improved a lot of things

### PR DESCRIPTION
**Pull request by Patitotective**
What the title says.
I should fix the conflicts...

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,40 +1,61 @@
 # Jitter
-A git-based binary manager for linux
+A git-based binary manager for linux written in Nim.
 
-## What is it?
-Jitter is a binary manager for linux. It searches github for executables that are avaliable to download. Unlike Homebrew or similar package managers, jitter does not require a brewfile or nixfile in order to recognize the project.
+## How it works?
+Jitter searches through GitHub for releases with `.tar.gz`, `.tgz` or `.zip` assets. Unlike Homebrew or similar package managers, jitter does not require a brewfile or nixfile in order to recognize the project.
 
+## Installing
```
Comment (ghost) [inline README.md:7]: good change, replacing the mug installer

File: README.md [hunk 2]
```diff
@@ -1,40 +1,61 @@
...
+- `jtr search vscodium`: searches and lists all repositories that have `vscodium` in their name.
+- `jtr search VSCodium/vscodium (tag|tags|true)`: searches and lists all release tags of repository `VSCodium/vscodium`
+- `jtr list`: lists all executables in Jitter's bin.
+- `jtr catalog`: lists all downloaded repositories
+- `jtr remove VSCodium/vscodium`: removes VSCodium/vscodium from your system.
+- `jtr install VSCodium/vscodium@1.69.0`: installs VSCodium/vscodium release with the tag `1.69.0`.
+- `jtr update VSCodium/vscodium`: updates vscodium to the latest version.
+- `jtr update (this|jitter|jtr)`: updates jitter to the latest release.
+
+## About
```
Comment (ghost) [inline README.md]: remove :)

File: README.md [hunk 3]
```diff
@@ -1,40 +1,61 @@
...
-10. `jtr update (this|jitter|jtr)` - updates jitter to the latest release
+Clone the repository and run `nimble build`.
+(You need to have nim and nimble installed).
+```
+git clone https://github.com/sharpcdf/jitter
+cd jitter
+nimble build
+```
+
+## Usage
```
Comment (ghost) [inline README.md:26]: you can just remove this, its the same info as ###examples

**Comment by ghost** [comment]
put parenthesis at the end of all procs, with or without args for source code consistency

File: jitter.nim [hunk 4]
```diff
@@ -1,237 +1,162 @@
...
+  let (ok, pkg) = input.parsePkg()
+  if not ok:
+    fatal fmt"Couldn't parse package {input}"
+
+  input.remove()
+  pkg.ghDownload(make)
+
+  success fmt"Successfully updated {pkg}"
+
+const parser = newParser:
```
Comment (ghost) [inline jitter.nim:115]: add doc comments to the arg parser

File: jitter/extract.nim [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import std/[strformat, strutils, osproc, os]
```
Comment (ghost) [inline jitter/extract.nim:1]: add a var for `.jitter/nerve` in files that use it a lot such as this

File: jitter/github.nim [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    content = client.getContent(url)
+  except HttpRequestError:
+    fatal "Failed to fetch repository"
+  finally:
+    client.close()
+
+  let data = content.parseJson()
+
+  if data.kind != JArray:
+    fatal fmt"Failed to fetch {pkg.gitFormat} releases"
```
Comment (ghost) [inline jitter/github.nim]: change fetch to find

File: jitter/github.nim [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+proc searchGhRepo*(repo: string): seq[Package] = 
+  let url = "https://api.github.com/search/repositories?" & encodeQuery({"q": repo})
+  let client = newHttpClient()
+  var content: string
+
+  try:
+    content = client.getContent(url)
+  except HttpRequestError:
+    fatal "Failed to fetch repositories"
```
Comment (ghost) [inline jitter/github.nim]: fetch to find :)

File: jitter/github.nim [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+  finally:
+    client.close()
+
+  let data = content.parseJson()
+  info &"Found {data[\"total_count\"]} repositories matching the query"
```
Comment (ghost) [inline jitter/github.nim]: remove this, no need

File: jitter/github.nim [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+  for repo in data["items"]:
+    list &"Github: {repo[\"full_name\"].getStr()}"
```
Comment (ghost) [inline jitter/github.nim]: remove `list`, just add and return and then echo it from the caller location

File: jitter/github.nim [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    result.add(parsePkg(repo["full_name"].getStr()).pkg)
+
+proc downloadRelease(pkg: Package, make = true) =
+  let url = 
+    if pkg.tag == "":
+      fmt"https://api.github.com/repos/{pkg.owner}/{pkg.repo}/releases/latest"
+    else:
+      fmt"https://api.github.com/repos/{pkg.owner}/{pkg.repo}/releases/tags/{pkg.tag}"
+
+  let client = newHttpClient(headers = newHttpHeaders([("accept", "application/vnd.github+json")]))
```
Comment (ghost) [inline jitter/github.nim:57]: make this the global client and remove others

File: jitter/github.nim [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    fatal fmt"No archives found for {pkg.gitFormat}"
+      
+  info fmt"Downloading {downloadUrl}"
+
+  client.downloadFile(downloadUrl, baseDir / "nerve" / downloadPath)
+  success fmt"Downloaded {pkg.gitFormat}"
+
+  extract(baseDir / "nerve" / downloadPath, pkg.dirFormat, make)
+
+proc ghDownload*(pkg: Package, make = true) =
```
Comment (ghost) [inline jitter/github.nim:105]: what is this????

File: jitter/github.nim [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+  if data.kind != JArray:
+    fatal fmt"Failed to fetch {pkg.gitFormat} releases"
+
+  info fmt"Listing release tags for {pkg.gitFormat}"
+
+  for release in data.getElems():
+    list release["tag_name"].getStr()
+    result.add(release["tag_name"].getStr())
+
+proc searchGhRepo*(repo: string): seq[Package] =
```
Comment (ghost) [inline jitter/github.nim]: change `searchGhRepo` to just `ghSearch`

File: jitter/github.nim [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+import extract, parse, log
+
+#TODO add support for appimage downloads
+#TODO prefer appimages -> .tar.gz -> .tgz -> .zip
+
+let baseDir = getHomeDir() & ".jitter/"
+let dDir = baseDir & "nerve/"
+
+proc listPkgGhReleases*(pkg: Package): seq[string] =
```
Comment (ghost) [inline jitter/github.nim]: change `listPkgGhReleases` to `ghListTags` or `ghListReleases`

**Review by ghost — COMMENTED** [review]
looks good, look at comments and make according changes :)

**Comment by ghost** [comment]
@Patitotective made a review

**Comment by ghost** [comment]
ill fix the conflicts afterwards

File: jitter/github.nim (hunk 10 above)
Comment (Patitotective) [reply jitter/github.nim:57]: That would create a client when `github.nim` is imported, that doesn't make any sense.

File: jitter/github.nim (hunk 11 above)
Comment (Patitotective) [reply jitter/github.nim:105]: It extracts the downloaded zip/tar to the package directory.

**Comment by ghost** [comment]
some called procs still have no parenthesis in jitter.nim - not a big deal, i can change with the conflict fixes

File: jitter.nim [hunk 14]
```diff
@@ -1,237 +1,166 @@
...
+    arg("query") ## Positional argument called query
+    run:
+      if opts.query.find(AllChars - IdentChars) >= 0: # If it finds more than IdentChars it should be a package and not a repo
+        let (ok, pkg) = opts.query.parsePkgFormat()
+        if not ok:
+          fatal fmt"Couldn't parse package {opts.query}"
+
+        discard pkg.ghListReleases()
+      else:
+        for pkg in opts.query.ghSearch():
```
Comment (ghost) [inline jitter.nim:150]: note: search for repos return empty names, tested with `jtr search vscodium`

File: jitter.nim [hunk 15]
```diff
@@ -1,237 +1,166 @@
...
+  removeDir(nerveDir / pkg.pkgFormat)
+
+proc remove(input: string) = 
+  let (ok, pkg) = input.parsePkgFormat()
+  if not ok:
+    fatal fmt"Couldn't parse package {input}"
+
+  let installedPkgs = getInstalledPkgs()
+
+  if not installedPkgs.anyIt(it.owner == pkg.owner and it.repo == pkg.repo):
```
Comment (ghost) [inline jitter.nim:67]: fails to locate package

**Comment by ghost** [comment]
i dont think you tested the changes after making them, i added the `config.nims` back, please make sure that the remove and other commands work first, i found search being inaccurate and remove not being able to find installed packages @Patitotective

File: jitter/extract.nim [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+import zippy/tarballs
+import zippy/ziparchives
+
+import log
+
+let baseDir = getHomeDir() & ".jitter/"
+let nerveDir = baseDir / "nerve"
+let binDir = baseDir / "bin"
```
Comment (ghost) [inline jitter/extract.nim:10]: add `bin/`, otherwise installs symlinks to parent directory with prefix `bin` (error)

Comment (ghost) [reply jitter/extract.nim:10]: might need to add same thing to `nerveDir`

**Comment by Patitotective** [comment]
> i dont think you tested the changes after making them, i added the `config.nims` back, please make sure that the remove and other commands work first, i found search being inaccurate and remove not being able to find installed packages @Patitotective

Well, I really didn't change any of the logic of the original code, I just improved the code itself
But sure, I'll test them

File: jitter/extract.nim (hunk 16 above)
Comment (Patitotective) [reply jitter/extract.nim:10]: That cannot happen because the `/` procedure removes trailing slashes so it's essentially the same.

Comment (Patitotective) [reply jitter/extract.nim:10]: That's not possible, the `/` proc removes trailing slashes.

**Pull request merged**